### ISSUE TITLE
Prerender in subprocess 

### ISSUE THREAD

**Issue by benmccann** [sveltejs/kit#5306]
### Describe the problem

SvelteKit `exit`s:

https://github.com/sveltejs/kit/blob/cfd38d40e035a6d9935d365acc2d079d9f15796a/packages/kit/src/vite/index.js#L312

This means you can't have another plugin after the SvelteKit plugin

### Describe the proposed solution

Add a new `close` hook

Have `prerender` call `server.close()`

### Importance

Prevents https://github.com/userquin/pwa-sveltekit-vite from working

**Comment by Rich-Harris** [comment]
Adding a new hook just for this purpose would be unfortunate, I think — especially since (as far as the user is concerned) the connection between prerendering and `close` is unclear. We'd also be relying on people to implement `close` correctly, otherwise their builds would never complete (and people would assume it was because of a SvelteKit bug).

Instead, if we want to avoid exiting the main process, we should probably just do the prerendering in a child process that we can safely kill once prerendering is complete.

**Comment by furudean** [comment]
A proper exit can't be guaranteed in all environments, especially serverless ones.

**Comment by Conduitry** [comment]
> A proper exit can't be guaranteed in all environments, especially serverless ones.

This issue is just referring to things that would happen during _prerendering_ right, and not at runtime? I would certainly hope that we can rely on all the Node stuff we want while the app is being built, regardless of which platform it is being deployed to.

**Comment by furudean** [comment]
Ack, my bad.

**Comment by userquin** [comment]
how about just adding a global `end` build hook, just call it with some context (Vite  and/or SvelteKit) before `process.exit`  is called when an adapter is configured (should also be called when the adapter is not configured)? This way you don't need to change anything, just adding the global end hook call before the exit and also when the adapter is not configured.

IIRC I sent a PR for it a few months ago when trying to include pwa plugin in SvelteKit using the rollup interplugin communication api.

**Comment by matiasfha** [comment]
Maybe completely wrong my questions here but, that this means that is not possible to run a plugin after the `closeBundle` hook. 
I'm looking for a way to run some transformation on the pre-renreder output files.

**Comment by benmccann** [comment]
@matiasfha the prerendered output files get generated in `writeBundle`, so I would expect you to be able to use `writeBundle` if your plugin is after the SvelteKit plugin. You may have to wait for https://github.com/sveltejs/kit/pull/5439 to be released first though to ensure that your changes would get copied by the adapters

If `exit` being called is preventing you from doing what you're trying to do, please feel free to send a PR to fix this issue. I don't know that we're likely to get to it anytime soon on our own since it doesn't affect most users

**Comment by userquin** [comment]
@benmccann I like the idea behind Astro and iles, where you have a lot of hooks, this way ppl can Just add their hooks in the plugin.

**Comment by matiasfha** [comment]
@benmccann the writeBundle hook can't read the prerendered output files in a reliable way. Looks like at the point the plugin calls this hook the previous plugin (the sveltekit one) is not yet done with it.
But, the adapt function called in closeBundle can do it.
If I write an adapter instead of a plugin the content of the files is available.
So, writing an adapter for this is not the way to go since the adapter is meant to do something different but would it be ok to add a new hook to be call before the adapter? I can do the PR but would like to check if it will be the correct path to allow users to run code before adapter to do some transformations.

**Comment by userquin** [comment]
@benmccann the plugin (there are 3) that generates the build has `enforce: 'post'`:
- https://github.com/antfu/vite-plugin-pwa/blob/v0.12.2/src/index.ts#L62
- https://github.com/antfu/vite-plugin-pwa/blob/v0.12.2/src/index.ts#L83 (this line will changed to `writeBundle` with the timeout via the sveltekit options).

**Comment by benmccann** [comment]
Hmm. I'm not sure why `enforce: 'post'` wouldn't work for either of you. The prerendering is `await`ed and doesn't happen in a worker, so I would expect it to be completely finished.

**Comment by dominikg** [comment]
@userquin  kit plugin does not set apply: https://github.com/sveltejs/kit/blob/578a2889363e6310d675d8207a5399f0ece52390/packages/kit/src/vite/index.js#L132

**Comment by userquin** [comment]
@dominikg uhmmm, the result is the same, confirmed with `patak`, so the problem is on the sveltekit plugin...

**Comment by dominikg** [comment]
yes, already added a PR to improve the kit plugin. Might be nice to split it into separate dev/build/common plugins to avoid conditionals inside the hooks :thinking:  but thats out of scope here. I wanted to mention it as you all were discussing behavior of closeBundle and it caught me by surprise today.

**Comment by matiasfha** [comment]
The prerendered process is in fact awaited here https://github.com/sveltejs/kit/blob/master/packages/kit/src/vite/index.js#L303 

but even with that a new plugin that hooks into using
- `closeBundle` nothings happens
- `writeBundle` or `buildEnd` the folder is not present yet

I'm testing with this little snippet

```
function buildPlugin(ctx) {
    
    return {
        name: 'vite-plugin-image-preprocess:build',
        enforce: 'post',
        apply: 'build',
        async writeBundle() {
            let filesContent = {}
            /**
             * The pre-rendered files will be produced and stored in the output folder
             * .svelte-kit/output/prerender/
             */
            debug(`Wait for pre-render process to finish`)
            const watcher = chokidar.watch('.svelte-kit/output').on('addDir', async (path, stats) => {
                if(path.includes('prerender')) {
                    debug(`Watcher ${path}`, stats)
					const files = fs.readdirSync('.svelte-kit/output')
			        debug(files)
                }
                
            })  
            await new Promise(resolve => setTimeout(resolve, 2000))            
        }
        
    }
}
```

This can find the folders `.svelte-kit/output/client` (and children) and `.svelte-kit/output/server`

This is running against a new sveltekit project from scratch with the simple todo example.

**Comment by benmccann** [comment]
This Vite folks confirmed that `enforce` only controls the order the plugins are started, but the parallel ones will all run in parallel still. We may need a new sequential hook in Vite for this
